core/GlobalState.cc

```diff
@@ -1412,6 +1418,72 @@ bool GlobalState::wasModified() const {
     return wasModified_;
 }
 
+bool GlobalState::isTypecheckingCanceled() const {
+    return lspEpochInvalidator->load() != currentlyProcessingLSPEpoch->load();
+}
+
+void GlobalState::startCommitEpoch(u4 epoch) {
+    absl::MutexLock lock(epochMutex.get());
+    currentlyProcessingLSPEpoch->store(epoch);
+    lspEpochInvalidator->store(epoch);
+    // These should always be different.
+    ENFORCE(epoch != lastCommittedLSPEpoch->load());
+}
+
+optional<pair<u4, u4>> GlobalState::getRunningSlowPath() const {
+    absl::MutexLock lock(epochMutex.get());
+    const u4 processing = currentlyProcessingLSPEpoch->load();
+    const u4 committed = lastCommittedLSPEpoch->load();
+    if (processing == committed) {
+        return nullopt;
+    }
+    return make_pair(committed, processing);
+}
+
+bool GlobalState::tryCancelSlowPath(u4 newEpoch) const {
+    absl::MutexLock lock(epochMutex.get());
+    const u4 processing = currentlyProcessingLSPEpoch->load();
+    ENFORCE(newEpoch != processing); // This would prevent a cancelation from happening.
+    const u4 committed = lastCommittedLSPEpoch->load();
+    // The second condition should never happen, but guard against it in production.
+    if (processing == committed || newEpoch == processing) {
+        return false;
+    }
+    // Cancel slow path by bumping invalidator.
+    lspEpochInvalidator->store(newEpoch);
+    return true;
+}
+
+bool GlobalState::tryCommitEpoch(u4 epoch, bool isCancelable, function<bool()> lambda) {
+    if (!isCancelable) {
+        return lambda();
+    }
+
+    // Should have called "startCommitEpoch" *before* this method.
+    ENFORCE(currentlyProcessingLSPEpoch->load() == epoch);
+    const bool canCommit = lambda();
+    {
+        absl::MutexLock lock(epochMutex.get());
+        if (canCommit) {
+            // Try to commit.
+            const u4 processing = currentlyProcessingLSPEpoch->load();
+            const u4 invalidator = lspEpochInvalidator->load();
+            if (processing == invalidator) {
+                ENFORCE(lastCommittedLSPEpoch->load() != processing, "Trying to commit an already-committed epoch.");
+                // OK to commit!
+                lastCommittedLSPEpoch->store(currentlyProcessingLSPEpoch->load());
```

Comment:
```suggestion
                lastCommittedLSPEpoch->store(processing);
```

Reply:
Alternatively, do you expect it to change? I'm confused what happens if it changes and if it could

Reply:
Good catch. I don't expect it to change.